src/botmation/factories/bot-actions-pipe.ts

```diff
@@ -0,0 +1,35 @@
+import { Page } from 'puppeteer'
+
+import { BotAction } from '../interfaces/bot-actions.interfaces'
+import { logError } from 'botmation/helpers/console'
+import { injectsPipeOrEmptyPipe, wrapValueInPipe } from 'botmation/helpers/pipe'
+
+//
+// Standards
+//  1) Pipes always inject a pipeable value at the end
+//  2) A piped value is always wrapped with branding, in case the number of injects is unknown
+//  3) A Bot action that does not return a value, in a pipe, effectively clears the piped value
+//  4) A cleared pipe is a piped value as undefined (so you get a piped branded object, with its `value` property = undefined)
+//  5) Pipes expect piped values to come in to their pipe, but have safe defaults
+//  7) Piped always return their piped values in the end (so if empty pipe, undefined, otherwise whatever the value, unbranded is)
+//
+export const BotActionsPipe = 
+  <R = any, P = any>(page: Page, ...injects: any[]) =>
+    async (...actions: BotAction[]): Promise<R> => {
+      // Possible for last inject to be the piped value
+      let pipe = injectsPipeOrEmptyPipe<P>(injects) // unwraps the piped value from the piped branded box
+
+      // let piped // pipe's are closed chain-links, so nothing pipeable comes in, so data is grabbed in a pipe and shared down stream a pipe, and returns
+      try {
+        for(const action of actions) {
+          const nextPipedValue = await action(page, ...injects, pipe) // if action doesn't return anything, is this value === undefined !?!?!?! TODO test this
+
+          pipe = wrapValueInPipe(nextPipedValue)
+        }
+      } catch(error) {
+        logError('PipeCaughtError:')
+        logError(error)
+      }
+
+      return pipe.value as any as R

```

Comment:
Returned piped wrapped value

Update `givenThat` to unbox piped value

That way sequences of action pipes can successfully pipe data through, in a pipe, so custom actions with pipes can interface properly

Reply:
Update:
ConditionalBotAction does NOT need to return a pipe, it can return a value instead, but it can return a value wrapped in a pipe as well

This allows ConditionalBotAction to use and return a BotActionsPipe, which always wraps the value on return

Utility actions were updated with `getPipeValue()` to check safely for value wrapped in a pipe or just value itself

So you can do it either way